src/main/scala/seek/aws/sns/SubscribeTopic.scala

```diff
@@ -34,34 +38,69 @@ class SubscribeTopic extends AwsTask {
       _  <- subscribeToTopics(ts, ps).run(c)
     } yield ()
 
+
   private def subscribeToTopics(ts: List[String], ps: List[PendingSubscription]): Kleisli[IO, AmazonSNS, Unit] =
     Kleisli { c =>
       ts.foldLeft(IO.unit) { (z, t) =>
         for {
           _  <- z
           es <- listSubscriptions(t).run(c).compile.toList
-          _  <- addSubscriptions(t, uniquePending(ps, es).toList).run(c)
+          _  <- updateSubscriptions(t, merge(ps, es)).run(c)
         } yield ()
     }
   }
 
-  private def addSubscriptions(topicArn: String, ps: List[PendingSubscription]): Kleisli[IO, AmazonSNS, Unit] =
+  private def updateSubscriptions(topicArn: String, ps: List[PendingSubscription]): Kleisli[IO, AmazonSNS, Unit] =
+    ps match {
+      case Nil    => Kleisli.liftF(IO.unit)
+      case h :: t => for {
+        arn <- maybeSubscribeToTopic(topicArn, h)
+        _   <- maybeUpdateFilterPolicy(arn, h.filterPolicy)
+        _   <- updateSubscriptions(topicArn, t)
+      } yield ()
+    }
+
+  private def maybeSubscribeToTopic(topicArn: String, ps: PendingSubscription): Kleisli[IO, AmazonSNS, String] =
+    Kleisli { c =>
+      ps.arn match {
+        case Some(a) => IO.pure(a)
+        case None => IO(c.subscribe(topicArn, ps.protocol, ps.endpoint).getSubscriptionArn)
+      }
+    }
+
+  private def maybeUpdateFilterPolicy(arn: String, policy: Option[String]): Kleisli[IO, AmazonSNS, Unit] =
     Kleisli { c =>
-      ps match {
-        case Nil    => IO.unit
-        case h :: t =>
-          for {
-            _ <- IO(c.subscribe(topicArn, h.protocol, h.endpoint))
-            _ <- addSubscriptions(topicArn, t).run(c)
-          } yield ()
+      policy match {
+        case Some(p) => IO {
+          val r = new SetSubscriptionAttributesRequest(arn, "FilterPolicy", p)
+          c.setSubscriptionAttributes(r)
```

Comment:
The `new SetSubscription...` shouldn't be inside the `IO`. Probably be simpler just to do do `c.setSubscriptionAttributes(new Set...`. Either way can you please align the `=>` by making the `None` come first?